### PR TITLE
Aden/searching

### PR DESCRIPTION
**Pull request by akrakman**

### CODE REVIEW

File: src/frontend/src/components/Searching.tsx [hunk 1]
```diff
@@ -1,20 +1,28 @@
...
-      pageNum
+      pageNum,
+      press,
+      selected
    );
-   console.log(apartments);
-
-   const [press, setPress] = useState(false);
-   //const [amountFound, setAmountFound] = useState(0);
+   //console.log(apartments);
```
Comment (MinhPhan8803) [inline src/frontend/src/components/Searching.tsx:25]: this line should be in testing only

File: src/frontend/src/components/useSearchApartment.tsx [hunk 2]
```diff
@@ -18,56 +23,113 @@ function useSearchBook(query: string, pageNum: number) {
...
-            for (let i = 0; i < res.data.length; i++) {
-               newApartments.push({
-                  name: res.data[i].name,
-                  address: res.data[i].address,
-                  image: res.data[i].image,
-                  review: res.data[i].review,
-                  rating: res.data[i].rating,
-                  price_min: res.data[i].price_min,
-                  price_max: res.data[i].price_max,
+      if (isInitialMount.current == true || query === '') {
```
Comment (MinhPhan8803) [inline src/frontend/src/components/useSearchApartment.tsx:30]: This one deals with populating the mainpage right?

File: src/frontend/src/components/Searching.tsx [hunk 3]
```diff
@@ -41,17 +49,20 @@ export default function Searching() {
...
+         handlePost(query, selected);
       }
    };
 
-   function handlePost(query: string) {
+   function handlePost(query: string, select: string[]) {
       axios({
-         method: 'post',
-         url: 'http://localhost:3333/mockdata', // need mainpage.py url
+         method: 'POST',
```
Comment (MinhPhan8803) [inline src/frontend/src/components/Searching.tsx:58]: This should be a get request, we are not posting stuff to the server.

File: src/frontend/src/components/Searching.tsx [hunk 4]
```diff
@@ -1,20 +1,28 @@
...
 import SingleCard from './SingleCard';
 import useSearchApartment from './useSearchApartment';
 
 export default function Searching() {
    const [query, setQuery] = useState('');
    const [pageNum, setPageNum] = useState(1);
+   const [press, setPress] = useState(false);
+   const emptyarray: string[] = [];
+   const [selected, setSelected] = useState(emptyarray);
    const { loading, error, apartments, hasMore } = useSearchApartment(
```
Comment (MinhPhan8803) [inline src/frontend/src/components/Searching.tsx:19]: Why is this called in `Searching.tsx`?

**Review by MinhPhan8803 — CHANGES_REQUESTED** [review]
Unless I understand the code wrong, I feel like there is some misunderstanding about the page design here. From the figma design, the mainpage would be populated separately from the search, which is based on a query string. The mainpage apartments on the other hand will be displayed directly on, well the mainpage, and the user can choose to sort these based on certain criterias. Hence, the search results and the displayed apartments are separate.

File: src/frontend/src/components/Searching.tsx (hunk 1 above)
Comment (akrakman) [reply src/frontend/src/components/Searching.tsx:25]: what line

**Comment by akrakman** [comment]
I'm not understanding what you mean. Sorry that we're still having this misunderstanding. Did you try this out locally? You can do `git fetch origin pull/34/head:aden/searching` and test it out. I'm pretty sure that's happening.

**Comment by akrakman** [comment]
I also neglected to put in comments. That is my fault.

File: src/frontend/src/components/Searching.tsx (hunk 3 above)
Comment (akrakman) [reply src/frontend/src/components/Searching.tsx:58]: How would the query get to the flask backend? The same thing happens in `Login.tsx`/`Register.tsx`.

**Comment by akrakman** [comment]
> > From the figma design, the mainpage would be populated separately from the search, which is based on a query string. 
How could the mainpage be populated without the query? If you mean before the query is put initially, then yes that happens and 10 appear when you load the page.

The mainpage apartments on the other hand will be displayed directly on, well the mainpage, and the user can choose to sort these based on certain criterias. Hence, the search results and the displayed apartments are separate.
> 
Oh, you mean I mean need to have the search results like in the figma. Okay, I didn't know how to do that so I went with this approach Possibly I could go back and do this.

> If these are already implemented separately then consider naming the files something more reflective of their purposes.
Will do.

**Comment by akrakman** [comment]
But in the Figma, is it the suggestions that appear on the left side or an option they pressed from the suggestions?

File: src/frontend/src/components/Searching.tsx (hunk 4 above)
Comment (akrakman) [reply src/frontend/src/components/Searching.tsx:19]: It receives the apartments from the GET requests, and whether they have more, is loading, etc.

File: src/frontend/src/components/useSearchApartment.tsx (hunk 2 above)
Comment (akrakman) [reply src/frontend/src/components/useSearchApartment.tsx:38]: I put in some clarifying comments.

**Comment by akrakman** [comment]
Should it be like this [https://lo-victoria.com/build-a-custom-autocomplete-search-bar-with-react-hooks](https://lo-victoria.com/build-a-custom-autocomplete-search-bar-with-react-hooks)?

File: src/frontend/src/components/Searching.tsx (hunk 3 above)
Comment (akrakman) [reply src/frontend/src/components/Searching.tsx:60]: If you mean getting all the apartments when the page loads, that would make the infinite scroll pretty hard

File: src/frontend/src/components/Searching.tsx (hunk 1 above)
Comment (MinhPhan8803) [reply src/frontend/src/components/Searching.tsx]: The log comment

> what line. deleted the log

File: src/frontend/src/components/Searching.tsx (hunk 3 above)
Comment (MinhPhan8803) [reply src/frontend/src/components/Searching.tsx:60]: We send a get request that contains some amount of apartments. The backend methods already took this into account.

> If you mean getting all the apartments when the page loads, that would make the infinite scroll pretty hard

Comment (MinhPhan8803) [reply src/frontend/src/components/Searching.tsx:60]: > How would the query get to the flask backend? The same thing happens in `Login.tsx`/`Register.tsx`. The flask backend then does a get request in the /home url.

We add a query string to the get request.

File: src/frontend/src/components/Searching.tsx (hunk 4 above)
Comment (MinhPhan8803) [reply src/frontend/src/components/Searching.tsx:19]: > It receives the apartments from the GET requests, and whether they have more, is loading, etc.

Search should only display apartment names though, the actual apartments are populated differently.

**Comment by MinhPhan8803** [comment]
> > From the figma design, the mainpage would be populated separately from the search, which is based on a query string.
> 
> How could the mainpage be populated without the query? If you mean before the query is put initially, then yes that happens and 10 appear when you load the page.
> 
> > The mainpage apartments on the other hand will be displayed directly on, well the mainpage, and the user can choose to sort these based on certain criterias. Hence, the search results and the displayed apartments are separate.
> 
> Oh, you mean I mean need to have the search results like in the figma. Okay, I didn't know how to do that so I went with this approach. Possibly I could go back and do this.
> 
> > If these are already implemented separately then consider naming the files something more reflective of their purposes.
> 
> Will do.

Yes, thats why we design the pages and the backend beforehand. So that each of us can have a concrete design to reference instead of figuring stuff out ourselves. As for the search, for now just make a pop up like in figma using the result from the request, we can deal with what happens when the user clicks on them later.

**Comment by MinhPhan8803** [comment]
> But in the Figma, do the suggestions appear on the left side or the option they pressed from the suggestions? Can you provide a sample query; maybe that will help me understand better. Or send a similar tutorial I could follow maybe

The search strings are displayed similar to what happens in google search or other social media searches: through a pop up.

**Comment by MinhPhan8803** [comment]
> > But in the Figma, do the suggestions appear on the left side or the option they pressed from the suggestions? Can you provide a sample query; maybe that will help me understand better. Or send a similar tutorial I could follow maybe
> 
> The search strings are displayed similar to what happens in google search or other social media searches: through a pop up.

This tutorial discusses how to create a search bar https://dev.to/salehmubashar/search-bar-in-react-js-545l

File: src/frontend/src/components/Searching.tsx (hunk 3 above)
Comment (MinhPhan8803) [reply src/frontend/src/components/Searching.tsx:60]: https://stackabuse.com/get-request-query-parameters-with-flask/
This discusses how the backend would handle the query strings.

Comment (akrakman) [reply src/frontend/src/components/Searching.tsx:60]: So I should do this thing [https://bobbyhadz.com/blog/react-update-query-params](https://bobbyhadz.com/blog/react-update-query-params)?

File: src/frontend/src/components/Searching.tsx (hunk 4 above)
Comment (akrakman) [reply src/frontend/src/components/Searching.tsx:19]: Screen Shot 2022-10-26 at 12 31 28 PM
Do are you saying I should have been doing the gray circle not the red circle?

Comment (akrakman) [reply src/frontend/src/components/Searching.tsx:19]: > The search strings are displayed similar to what happens in google search or other social media searches: through a pop up.
>
Screen Shot 2022-10-26 at 12 33 34 PM

Is this what you mean by pop up?

**Comment by akrakman** [comment]
> The search strings are displayed similar to what happens in google search or other social media searches: through a pop up.
>
Screen Shot 2022-10-26 at 12 33 34 PM

Is this what you mean by pop up?

**Comment by MinhPhan8803** [comment]
> > The search strings are displayed similar to what happens in google search or other social media searches: through a pop up.
> 
> Screen Shot 2022-10-26 at 12 33 34 PM
> 
> Is this what you mean by pop up?

Yes

**Comment by akrakman** [comment]
Wait there are some other comment replies above also.

File: src/frontend/src/components/Searching.tsx (hunk 4 above)
Comment (MinhPhan8803) [reply src/frontend/src/components/Searching.tsx:19]: The `search` thing in mainpage corresponds to the grey circle, similar to what I explained below. This corresponds to the `search_apartments()` method in `Mainpage.py`.

The list of apartments are displayed by default on the page, which the user can sort based on price and rating. This corresponds to the various `apartments_*.py` methods in `Mainpage.py`.

So yes, search should be the grey circle, the red circle is different.

**Comment by MinhPhan8803** [comment]
Remember to fix the eslint warnings. The action treats all warnings as errors.

**Comment by MinhPhan8803** [comment]
I cant look at the branch rn, can you post screenshots of what it currently look like?

File: src/frontend/src/components/getApartments.tsx [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+function getApartments(query: string, pageNum: number, selected: string[]) {
+   const [loading, setLoading] = useState(true);
+   const [error, setError] = useState(false);
+   const array: {
+      // avoids linting warnings
+      name: string;
+      address: string;
+      image: string;
+      review: string;
```
Comment (MinhPhan8803) [inline src/frontend/src/components/getApartments.tsx:12]: The review(s) ans image(s) are handled by the right half, this list of apartments only need to store what is in the figma design.

File: src/frontend/src/components/getApartments.tsx [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+               rating: number;
+               price_min: string;
+               price_max: string;
+            }[] = [];
+            for (let i = 0; i < res.data.length; i++) {
+               if (res.data[i].name !== undefined) {
+                  newApartments.push({
+                     name: res.data[i].name,
+                     address: res.data[i].address,
+                     image: res.data[i].image,
```
Comment (MinhPhan8803) [inline src/frontend/src/components/getApartments.tsx:56]: Similarly here, first thing is that there are multiple images and reviews that will be handled by the right half. Second thing is that the jsons sent by this pull request will not contain them per the Figma design and the Apartments class in the components spreadsheet.

File: src/frontend/src/components/SearchAndDisplayResults.tsx [hunk 7]
```diff
@@ -33,52 +41,25 @@ export default function Searching() {
...
-      setPageNum(1);
-      setPress(false);
-   };
-
-   const handlePress = (e: KeyboardEvent<HTMLInputElement>) => {
-      // POST request when enter is pressed
-      if (e.key === 'Enter') {
-         e.preventDefault();
-         setPress(true);
+   const handleSubmit = (
```
Comment (MinhPhan8803) [inline src/frontend/src/components/SearchAndDisplayResults.tsx:45]: Search is not the same thing as getting apartments to display.

**Review by MinhPhan8803 — CHANGES_REQUESTED** [review]
Can we hop on a call some time later tonight, I feel like something needs to be cleared up here.

File: src/frontend/src/components/getApartments.tsx (hunk 6 above)
Comment (MinhPhan8803) [reply src/frontend/src/components/getApartments.tsx:56]: > Similarly here, first thing is that there are multiple images and reviews that will be handled by the right half. Second thing is that the jsons sent by the backend for this task will not contain them per the Figma design and the Apartments class in the components spreadsheet.

File: src/frontend/src/components/getApartments.tsx (hunk 5 above)
Comment (akrakman) [reply src/frontend/src/components/getApartments.tsx:12]: I know the review will be removed but we still need to image

File: src/frontend/src/components/getApartments.tsx (hunk 6 above)
Comment (akrakman) [reply src/frontend/src/components/getApartments.tsx:56]: yeah I didn't remove it yet

**Comment by akrakman** [comment]
> Can we hop on a call some time later tonight, I feel like something needs to be cleared up here.

Call me asap

File: src/frontend/src/components/SearchAndDisplayResults.tsx (hunk 7 above)
Comment (akrakman) [reply src/frontend/src/components/SearchAndDisplayResults.tsx:45]: explain

File: src/frontend/src/components/getApartments.tsx (hunk 6 above)
Comment (akrakman) [reply src/frontend/src/components/getApartments.tsx]: removed

File: src/frontend/src/components/getApartments.tsx (hunk 5 above)
Comment (akrakman) [reply src/frontend/src/components/getApartments.tsx]: so you're saying I should store the images in a different object?

File: src/backend/requirements.txt [hunk 8]
```diff
@@ -2,4 +2,6 @@ Flask==2.2.2
 pytest==7.1.3
 pylint==2.15.2
 coverage==6.4.4
-black==22.8.0
\ No newline at end of file
+black==22.8.0
+marshmallow==3.14.1
```
Comment (MinhPhan8803) [inline src/backend/requirements.txt:6]: remove marshmallow and cors, we are not using them yet

File: src/frontend/src/components/populateLeft.tsx [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+      setSelected(newSelected);
+      // sets URL
+      if (newSelected.includes('low-high')) {
+         searchParams.set('priceSort', '-1');
+      } else if (newSelected.includes('high-low')) {
+         searchParams.set('priceSort', '1');
+      } else {
+         searchParams.delete('priceSort');
+      }
+      if (newSelected.includes('most popular')) {
```
Comment (MinhPhan8803) [inline src/frontend/src/components/populateLeft.tsx:63]: there is also a "least popular" option

**Review by MinhPhan8803 — CHANGES_REQUESTED** [review]

**Comment by akrakman** [comment]
@MinhPhan8803  do you approve

**Comment by MinhPhan8803** [comment]
> @MinhPhan8803 do you approve

Need to solve the conflicts first.

**Review by MinhPhan8803 — APPROVED** [review]

**Pull request merged**